src/cascadia/TerminalControl/TermControl.cpp

```diff
@@ -2526,16 +2527,57 @@ namespace winrt::Microsoft::Terminal::Control::implementation
 
             if (items.Size() > 0)
             {
-                std::wstring allPaths;
-                for (auto item : items)
+                std::vector<std::wstring> fullPaths;
+
+                // GH#14628: Workaround for GetStorageItemsAsync() only returning 16 items
+                // at most when dragging and dropping from archives (zip, 7z, rar, etc.)
+                if (items.Size() == 16 && e.DataView().Contains(winrt::hstring{ L"FileDrop" }))
                 {
-                    // Join the paths with spaces
-                    if (!allPaths.empty())
+                    auto fileDropData = co_await e.DataView().GetDataAsync(winrt::hstring{ L"FileDrop" });
+                    if (fileDropData != nullptr)
+                    {
+                        auto stream = fileDropData.as<IRandomAccessStream>();
+                        stream.Seek(0);
+
+                        const uint32_t streamSize = gsl::narrow_cast<uint32_t>(stream.Size());
+                        const Buffer buf(streamSize);
+                        const auto buffer = co_await stream.ReadAsync(buf, streamSize, InputStreamOptions::None);
+
+                        const HGLOBAL hGlobal = buffer.data();
+                        const auto count = DragQueryFileW(static_cast<HDROP>(hGlobal), 0xFFFFFFFF, nullptr, 0);
+                        fullPaths.reserve(count);
+
+                        for (unsigned int i = 0; i < count; i++)
+                        {
+                            std::wstring path;
+                            path.resize(wil::max_path_length);

```

Comment:
`wil::max_path_length` is 259, it doesn't include the null terminator
maybe a potential issue for files with max name lengths?

Reply:
Ah, so basic_string has one CharT-sized unit at the end of its allocated buffer that is safe to overwrite with `\0`. That should give us the 260 we need, right?

I almost switched over to `wil::AdaptFixedSizeToAllocatedResult`, but a quick cross reference with the implementation of DragFileQuery revealed that we would never need anything more than MAX_PATH

Reply:
oh cool! that will definitely work then.
if we wanted to be more optimal, `DragQueryFile` returns the exact chars needed if you pass in `null` as the buffer.

lmk if this is something we want to do before we merge
```cpp
  for (unsigned int i = 0; i < count; i++)
  {
      const auto charsNeeded = DragQueryFileW(static_cast<HDROP>(hGlobal), i, nullptr, 0);
      std::wstring path;
      path.resize(charsNeeded);
      const auto charsCopied = DragQueryFileW(static_cast<HDROP>(hGlobal), i, path.data(), path.capacity());

      if (charsCopied > 0)
      {
          path.resize(charsCopied);
          fullPaths.emplace_back(std::move(path));
      }
  }
```

Reply:
I'm cool with keeping a 260-char buffer somewhere; I know I moved it from the stack to (likely) the heap, but honestly... eh.